File: Clion/ZeroTrac/Tree/isSubtree_572.cpp
//
// Created by TzashiNorpu on 11/3/2022.
//
#include "tree.h"

using namespace Tree;

bool Solution::isSubtree(TreeNode *root, TreeNode *subRoot) {
  function<bool(TreeNode *, TreeNode *)> helper = [&](TreeNode *node1, TreeNode *node2) -> bool {
      if (!node1 && !node2) return true;
      if (!node1 || !node2) return false;
      if (node2->val == node1->val)
        return helper(node1->left, node2->left) && helper(node1->right, node2->right);
      return false;
  };
  if (!root) return false;
  if (helper(root, subRoot)) return true;
  return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
}

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/isSubtree_572.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class isSubtree_572 {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null) return false;
        if (isSame(root, subRoot)) return true;
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    private boolean isSame(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) return true;
        if (root == null || subRoot == null) return false;
        if (root.val == subRoot.val) return isSame(root.left, subRoot.left) && isSame(root.right, subRoot.right);
        return false;
    }
}
